src/Winton.Extensions.Configuration.Consul/ConsulConfigurationProvider.cs

```diff
@@ -50,8 +49,8 @@ public void Dispose()
                 return;
             }
 
-            _cancellationTokenSource.Cancel();
-            _cancellationTokenSource.Dispose();
+            _source.WatchCancellationTokenSource!.Cancel();
+            _source.WatchCancellationTokenSource!.Dispose();
```

Comment:
Having thought about this change (which is the one I had envisioned making myself from in my previous comment on the issue), I remember now why I made this `CancelationTokenSource` private in this class. 

The problem we have with this new design is that it's no longer clear who the resource owner is which makes this `Dispose` implementation technically incorrect because if the token is created outside of here then it may well get `Disposed` before this call. 

A "trivial" workaround would be the check if this token has already been cancelled or something, but I think it's really just pointing to a bigger design problem here or multiple resource ownership. 

Instead, if we want to make it possible for external clients to cancel the `PollingLoop` (which is reasonable still I think) then we should keep two separate `CancelationTokenSource` instances. One which is internal which is used to shut down the `PollingLoop` when this class is `Disposed` (as it currently does). The second being one that is optionally set on the `ConsulConfigurationSource` which is purely owned by the client if they decide to set it. Such that we never have to worry about disposing of it in here. 

What do you think?

Reply:
Indeed, your concerns also seemed reasonable to me. Taking that into consideration, I will make the update as shown below, if you give the green light.

```
    internal sealed class ConsulConfigurationProvider : ConfigurationProvider, IDisposable
    {
        private readonly CancellationTokenSource _cancellationTokenSource;
        private readonly IConsulClientFactory _consulClientFactory;
        private readonly IConsulConfigurationSource _source;
        private ulong _lastIndex;
        private Task? _pollTask;
        private bool _disposed;

        public ConsulConfigurationProvider(
            IConsulConfigurationSource source,
            IConsulClientFactory consulClientFactory)
        {
            if (source.Parser == null)
            {
                throw new ArgumentNullException(nameof(source.Parser));
            }

            _source = source;
            _consulClientFactory = consulClientFactory;
            _cancellationTokenSource = new CancellationTokenSource();
            if (_source.WatchCancellationTokenSource != null)
            {
                _source.WatchCancellationTokenSource.Token.Register(() =>
                {
                    if (!_disposed && !_cancellationTokenSource.IsCancellationRequested)
                    {
                        _cancellationTokenSource.Cancel();
                    }
                });
            }
        }
```

Reply:
Yeah nice that's looks good to me. Thanks.

Reply:
I updated the pr, thanks.